### ISSUE TITLE
Long-running redis `KEYS` command caused external MemoryDB storage to hang

### ISSUE THREAD

**Issue by edoakes** [ray-project/ray#32537]
Ray version: 2.2.0

I am running a number of Ray clusters to run Ray Serve applications to test the failure handling for GCS fault tolerance with external storage.

I'm using am AWS MemoryDB instance (one node, one shard) in the same VPC as the cluster instances for external GCS storage.

MemoryDB had been stable for a few weeks (with an average of ~10s of clusters running against it) until yesterday when it started rejecting connections and all head nodes started to crash while connecting.

I filed an AWS support ticket and they dug into the issue. Their response was that there seemed to be a long-running `KEYS` command causing the single-threaded redis engine to be blocked for an extended period of time (their response is attached as a PDF).

Below are some screenshots of the MemoryDB cloudwatch metrics during the incident:

<img width="792" alt="Screen Shot 2023-02-14 at 12 04 02 PM" src="https://user-images.githubusercontent.com/9871461/218821817-451b7d68-9066-4f14-b418-10ca2304472c.png">
<img width="539" alt="Screen Shot 2023-02-14 at 12 04 14 PM" src="https://user-images.githubusercontent.com/9871461/218821818-8afc696e-d326-407c-8fa0-0b9005112348.png">

And the response from AWS support:
[memorydb_hang_support_response.pdf](https://github.com/ray-project/ray/files/10735537/memorydb_hang_support_response.pdf)

**Comment by edoakes** [comment]
Output of `INFO` on the offending redis cluster:
```
eoakes-test-no-cluster.supuit.clustercfg.memorydb.us-east-2.amazonaws.com:6379> INFO
# Server
redis_version:6.2.6
redis_git_sha1:0
redis_git_dirty:0
redis_build_id:0
redis_mode:cluster
os:Amazon MemoryDB
arch_bits:64
multiplexing_api:epoll
atomicvar_api:c11-builtin
gcc_version:0.0.0
process_id:1
run_id:0f873d5d94e4fb2e6fb1186e6d543f922a65f851
tcp_port:6379
server_time_usec:1676398678406883
uptime_in_seconds:63373
uptime_in_days:0
hz:10
configured_hz:10
lru_clock:15454294
executable:-
config_file:-

# Clients
connected_clients:8
cluster_connections:0
maxclients:65000
client_recent_max_input_buffer:64
client_recent_max_output_buffer:0
blocked_clients:0
tracking_clients:0
clients_in_timeout_table:0

# Memory
used_memory:393294552
used_memory_human:375.07M
used_memory_rss:444608512
used_memory_rss_human:424.01M
used_memory_peak:740813208
used_memory_peak_human:706.49M
used_memory_peak_perc:53.09%
used_memory_overhead:29076152
used_memory_startup:10209848
used_memory_dataset:364218400
used_memory_dataset_perc:95.08%
allocator_allocated:405073232
allocator_active:414584832
allocator_resident:426668032
used_memory_lua:32768
used_memory_lua_human:32.00K
used_memory_scripts:0
used_memory_scripts_human:0B
number_of_cached_scripts:0
maxmemory:14037181030
maxmemory_human:13.07G
maxmemory_policy:noeviction
allocator_frag_ratio:1.02
allocator_frag_bytes:9511600
allocator_rss_ratio:1.03
allocator_rss_bytes:12083200
rss_overhead_ratio:1.04
rss_overhead_bytes:17940480
mem_fragmentation_ratio:1.13
mem_fragmentation_bytes:51315512
mem_not_counted_for_evict:0
mem_replication_backlog:1048576
mem_clients_slaves:20512
mem_clients_normal:164080
mem_aof_buffer:0
mem_allocator:jemalloc-5.1.0
active_defrag_running:0
lazyfree_pending_objects:0
lazyfreed_objects:0

# Persistence
loading:0
current_cow_size:0
current_cow_size_age:0
current_fork_perc:0.00
current_save_keys_processed:0
current_save_keys_total:0
rdb_changes_since_last_save:2397898
rdb_bgsave_in_progress:0
rdb_last_save_time:1676335305
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:-1
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:0
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0
module_fork_in_progress:0
module_fork_last_cow_size:0

# SSL
ssl_enabled:no
ssl_current_certificate_not_before_date:(null)
ssl_current_certificate_not_after_date:(null)
ssl_current_certificate_serial:(null)

# Stats
total_connections_received:292451
total_commands_processed:41544415
instantaneous_ops_per_sec:9
total_net_input_bytes:2255157520
total_net_output_bytes:1600801869
instantaneous_input_kbps:0.51
instantaneous_output_kbps:20.19
rejected_connections:0
sync_full:0
sync_partial_ok:1
sync_partial_err:0
expired_keys:0
expired_stale_perc:0.00
expired_time_cap_reached_count:0
expire_cycle_cpu_milliseconds:828
evicted_keys:0
keyspace_hits:32834
keyspace_misses:244
pubsub_channels:0
pubsub_patterns:0
latest_fork_usec:0
total_forks:0
migrate_cached_sockets:0
slave_expires_tracked_keys:0
active_defrag_hits:0
active_defrag_misses:0
active_defrag_key_hits:0
active_defrag_key_misses:0
tracking_total_keys:0
tracking_total_items:0
tracking_total_prefixes:0
unexpected_error_replies:0
total_error_replies:5826
dump_payload_sanitizations:0
total_reads_processed:33667466
total_writes_processed:32562971
io_threaded_reads_processed:0
io_threaded_writes_processed:0

# Replication
role:master
connected_slaves:0
master_failover_state:no-failover
master_replid:edc6c6852b32208287ec044f246934defa8a8edf
master_replid2:bd5ebc78c20be99ace9af6e24819afcf57c59f48
master_repl_offset:10665722915
second_repl_offset:10652564999
repl_backlog_active:1
repl_backlog_size:1048576
repl_backlog_first_byte_offset:10664674340
repl_backlog_histlen:1048576

# CPU
used_cpu_sys:982.682434
used_cpu_user:4170.007127
used_cpu_sys_children:0.000000
used_cpu_user_children:0.001253
used_cpu_sys_main_thread:371.677184
used_cpu_user_main_thread:1236.847637

# Modules

# Errorstats
errorstat_CLUSTERDOWN:count=3000
errorstat_CROSSSLOT:count=338
errorstat_ERR:count=2
errorstat_LOADING:count=2486

# Cluster
cluster_enabled:1

# Keyspace
db0:keys=335970,expires=0,avg_ttl=0
```

**Comment by edoakes** [comment]
Output of `MEMORY STATS` on the offending cluster:
```
eoakes-test-no-cluster.supuit.clustercfg.memorydb.us-east-2.amazonaws.com:6379> MEMORY STATS
 1) "peak.allocated"
 2) (integer) 740813208
 3) "total.allocated"
 4) (integer) 393462344
 5) "startup.allocated"
 6) (integer) 10209848
 7) "replication.backlog"
 8) (integer) 1048576
 9) "clients.slaves"
10) (integer) 20512
11) "clients.normal"
12) (integer) 246064
13) "aof.buffer"
14) (integer) 0
15) "lua.caches"
16) (integer) 0
17) "db.0"
18) 1) "overhead.hashtable.main"
    2) (integer) 17633104
    3) "overhead.hashtable.expires"
    4) (integer) 32
19) "overhead.total"
20) (integer) 29158136
21) "keys.count"
22) (integer) 335970
23) "keys.bytes-per-key"
24) (integer) 1140
25) "dataset.bytes"
26) (integer) 364304208
27) "dataset.percentage"
28) "95.055923461914062"
29) "peak.percentage"
30) "53.112220764160156"
31) "allocator.allocated"
32) (integer) 405601512
33) "allocator.active"
34) (integer) 415158272
35) "allocator.resident"
36) (integer) 426901504
37) "allocator-fragmentation.ratio"
38) "1.023561954498291"
39) "allocator-fragmentation.bytes"
40) (integer) 9556760
41) "allocator-rss.ratio"
42) "1.0282860994338989"
43) "allocator-rss.bytes"
44) (integer) 11743232
45) "rss-overhead.ratio"
46) "1.0433393716812134"
47) "rss-overhead.bytes"
48) (integer) 18501632
49) "fragmentation"
50) "1.1318919658660889"
51) "fragmentation.bytes"
52) (integer) 51899912
```

**Comment by edoakes** [comment]
Output of `SLOWLOG` below. The longest was ~45000 microseconds (45ms). Seems unlikely that there was actually a long command blocking the thread.

```
internal-testing.supuit.clustercfg.memorydb.us-east-2.amazonaws.com:6379> SLOWLOG GET
 1) 1) (integer) 59527
    2) (integer) 1676487916
    3) (integer) 11057
    4) 1) "SCAN"
       2) "54595"
       3) "MATCH"
       4) "ses_idennjwx8w7sj7x1428jg75l7y@JOB:*"
       5) "COUNT"
       6) "1000"
    5) "10.1.107.238:46782"
    6) ""
 2) 1) (integer) 59526
    2) (integer) 1676487716
    3) (integer) 15605
    4) 1) "SCAN"
       2) "189923"
       3) "MATCH"
       4) "ses_idennjwx8w7sj7x1428jg75l7y@JOB:*"
       5) "COUNT"
       6) "1000"
    5) "10.1.107.238:46782"
    6) ""
 3) 1) (integer) 59525
    2) (integer) 1676487651
    3) (integer) 12674
    4) 1) "SCAN"
       2) "238107"
       3) "MATCH"
       4) "ses_idennjwx8w7sj7x1428jg75l7y@JOB:*"
       5) "COUNT"
       6) "1000"
    5) "10.1.107.238:46782"
    6) ""
 4) 1) (integer) 59524
    2) (integer) 1676487574
    3) (integer) 10616
    4) 1) "SCAN"
       2) "143896"
       3) "MATCH"
       4) "ses_idennjwx8w7sj7x1428jg75l7y@JOB:*"
       5) "COUNT"
       6) "1000"
    5) "10.1.107.238:46782"
    6) ""
 5) 1) (integer) 59523
    2) (integer) 1676487525
    3) (integer) 10681
    4) 1) "SCAN"
       2) "150611"
       3) "MATCH"
       4) "ses_e49barzkwpt2kk2dfjztfevu2b@JOB:*"
       5) "COUNT"
       6) "1000"
    5) "10.1.116.193:40738"
    6) ""
 6) 1) (integer) 59522
    2) (integer) 1676487525
    3) (integer) 10624
    4) 1) "SCAN"
       2) "124067"
       3) "MATCH"
       4) "ses_e49barzkwpt2kk2dfjztfevu2b@JOB:*"
       5) "COUNT"
       6) "1000"
    5) "10.1.116.193:40738"
    6) ""
 7) 1) (integer) 59521
    2) (integer) 1676487480
    3) (integer) 15823
    4) 1) "SCAN"
       2) "189285"
       3) "MATCH"
       4) "ses_e49barzkwpt2kk2dfjztfevu2b@JOB:*"
       5) "COUNT"
       6) "1000"
    5) "10.1.116.193:40738"
    6) ""
 8) 1) (integer) 59520
    2) (integer) 1676487458
    3) (integer) 48686
    4) 1) "KEYS"
       2) "ses_e49barzkwpt2kk2dfjztfevu2b@@namespace_job:_ray_internal_job_info_*"
    5) "10.1.116.193:40758"
    6) ""
 9) 1) (integer) 59519
    2) (integer) 1676487458
    3) (integer) 12990
    4) 1) "SCAN"
       2) "223791"
       3) "MATCH"
       4) "ses_e49barzkwpt2kk2dfjztfevu2b@JOB:*"
       5) "COUNT"
       6) "1000"
    5) "10.1.116.193:40738"
    6) ""
10) 1) (integer) 59518
    2) (integer) 1676487456
    3) (integer) 45509
    4) 1) "KEYS"
       2) "ses_e49barzkwpt2kk2dfjztfevu2b@@namespace_job:_ray_internal_job_info_*"
    5) "10.1.116.193:40758"
    6) ""
```

**Comment by cadedaniel** [comment]
@edoakes and @iycheng confirmed that this has been fixed in 2.3. It is no longer a release blocker.

**Comment by edoakes** [comment]
Have been continuously testing for the past few days without any issue on the `ray==2.3.0` RC, so this appears to be fixed.
